File: projects/java/cdm/src/mil/tatrc/physiology/datamodel/patient/actions/SEMechanicalVentilation.java
/**************************************************************************************
Copyright 2015 Applied Research Associates, Inc.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the License
at:
http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
 **************************************************************************************/

package mil.tatrc.physiology.datamodel.patient.actions;

import java.util.ArrayList;
import java.util.List;

import org.jfree.util.Log;

import mil.tatrc.physiology.datamodel.CDMSerializer;
import mil.tatrc.physiology.datamodel.bind.EnumOnOff;
import mil.tatrc.physiology.datamodel.bind.EnumSubstanceState;
import mil.tatrc.physiology.datamodel.bind.MechanicalVentilationData;
import mil.tatrc.physiology.datamodel.bind.SubstanceFractionData;
import mil.tatrc.physiology.datamodel.properties.*;
import mil.tatrc.physiology.datamodel.substance.SESubstance;
import mil.tatrc.physiology.datamodel.substance.SESubstanceFraction;
import mil.tatrc.physiology.datamodel.substance.SESubstanceManager;
import mil.tatrc.physiology.datamodel.system.SESystem;
import mil.tatrc.physiology.utilities.DoubleUtils;

public class SEMechanicalVentilation extends SEPatientAction
{
  protected SEScalarVolumePerTime flow;
  protected SEScalarPressure      pressure;
  protected EnumOnOff             state;

	protected List<SESubstanceFraction>  gasFractions;

  public SEMechanicalVentilation()
  {
    clear();
  }
  
  public void clear()
  {
  	flow = null;
  	pressure = null;
  	state = null;
		this.gasFractions=new ArrayList<SESubstanceFraction>();
  }

  public void reset()
  {
    state = null;
    if (flow != null)
    	flow.invalidate();
    if (pressure != null)
    	pressure.invalidate();
    this.gasFractions.clear();
  }

  public void copy(SEMechanicalVentilation from)
  {
    reset();

    setState(from.getState());
    if(from.hasFlow())
    	getFlow().set(from.getFlow());
    if(from.hasPressure())
    	getPressure().set(from.getPressure());    
    
    if(from.gasFractions!=null)
    {
      SESubstanceFraction mine;
      for(SESubstanceFraction sf : from.gasFractions)
      {
        mine=this.createGasFraction(sf.getSubstance());
        if(sf.hasFractionAmount())
          mine.getFractionAmount().set(sf.getFractionAmount());
      }
    }    
  }
  
  @Override
	public boolean isValid()
	{
  	if (!hasState())
    {
      Log.error("Mechanical Ventilation must have state.");
      return false;
    }
    if (getState() == EnumOnOff.OFF)
      return true;
    if (!hasGasFraction())
    {
    	Log.error("Mechanical Ventilation must have gas fractions.");
      return false;
    }
    else
    {
      double total = 0;
      for (SESubstanceFraction sf : gasFractions)
      {
        total += sf.getFractionAmount().getValue();
      }
      if (!DoubleUtils.equals(1.0, total))
      {
      	Log.error("Mechanical Ventilation Gas fractions do not sum to 1");
        return false;
      }
    }
    if (!hasPressure() && !hasFlow())
    {
    	Log.error("Mechanical Ventilation must have pressure and/or flow");
      return false;
    }
    return true;
	}

  public boolean load(MechanicalVentilationData in, SESubstanceManager substances)
  {
  	this.reset();
    if (in.getState() != null)
      setState(in.getState());
    if (in.getFlow() != null)
    	getFlow().load(in.getFlow());
    if (in.getPressure() != null)
    	getPressure().load(in.getPressure());
    
    SESubstance sub;
		if(in.getGasFraction()!=null)
		{
			for(SubstanceFractionData subData : in.getGasFraction())
			{
				sub = substances.getSubstance(subData.getName());
				if(sub == null)
				{
					Log.error("Substance does not exist : "+subData.getName());
					return false;
				}
				if(sub.getState() != EnumSubstanceState.GAS)
				{
					Log.error("Gas Fraction substance must be a gas, "+subData.getName()+" is not a gas...");
					return false;
				}
				this.createGasFraction(sub).getFractionAmount().load(subData.getFractionAmount());
			}
		}
    return isValid();
  }

  public MechanicalVentilationData unload()
  {
  	MechanicalVentilationData data = CDMSerializer.objFactory.createMechanicalVentilationData();
    unload(data);
    return data;
  }

  protected void unload(MechanicalVentilationData data)
  {
    if (hasState())
      data.setState(state);
    if (hasFlow())
      data.setFlow(flow.unload());
    if (hasPressure())
      data.setPressure(pressure.unload());
    
    for(SESubstanceFraction gf : this.gasFractions)
			data.getGasFraction().add(gf.unload());
  }

  public EnumOnOff getState()
  {
    return state;
  }
  public void setState(EnumOnOff state)
  {
    this.state = state;
  }
  public boolean hasState()
  {
    return state == null ? false : true;
  }

  public boolean hasFlow()
  {
    return flow == null ? false : flow.isValid();
  }
  public SEScalarVolumePerTime getFlow()
  {
    if (flow == null)
    	flow = new SEScalarVolumePerTime();
    return flow;
  }

  public boolean hasPressure()
  {
    return pressure == null ? false : pressure.isValid();
  }
  public SEScalarPressure getPressure()
  {
    if (pressure == null)
    	pressure = new SEScalarPressure();
    return pressure;
  }
  
  public SESubstanceFraction createGasFraction(SESubstance substance)
	{
		return getGasFraction(substance);
	}
	public SESubstanceFraction getGasFraction(SESubstance substance)
	{
		for(SESubstanceFraction sf : this.gasFractions)
		{
			if(sf.getSubstance().getName().equals(substance.getName()))
			{				
				return sf;
			}
		}		
		SESubstanceFraction sf = new SESubstanceFraction(substance);		
		this.gasFractions.add(sf);
		return sf;
	}
	public boolean hasGasFraction()
	{
		return !this.gasFractions.isEmpty();
	}
	public boolean hasGasFraction(SESubstance substance)
	{
		for(SESubstanceFraction sf : this.gasFractions)
		{
			if(sf.getSubstance()==substance)
			{				
				return true;
			}
		}
		return false;
	}
	public List<SESubstanceFraction> getGasFraction()
	{
		return this.gasFractions;
	}
	public void removeGasFraction(SESubstance substance)
	{
		for(SESubstanceFraction sf : this.gasFractions)
		{
			if(sf.getSubstance()==substance)
			{
				this.gasFractions.remove(sf);
				return;
			}
		}	
	}

	public String toString()
	{
		String cnts = "Mechanical Ventilation"
				+ "\n\tState: " + getState()
		    + "\n\tFlow: " + (hasFlow() ? getFlow() : "Not Provided")
		    + "\n\tPressure: " + (hasPressure() ? getPressure() : "Not Provided");
		for(SESubstanceFraction sf : this.gasFractions)
			cnts += "\n\tSubstanceFraction: " + sf.getSubstance().getName() + " : " + sf.getFractionAmount();
		
		return cnts;
	}
}

File: projects/biogears/libBiogears/src/cdm/patient/actions/SEMechanicalVentilation.cpp
/**************************************************************************************
Copyright 2015 Applied Research Associates, Inc.
Licensed under the Apache License, Version 2.0 (the "License"); you may not use
this file except in compliance with the License. You may obtain a copy of the License
at:
http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software distributed under
the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
**************************************************************************************/
#include <biogears/cdm/patient/actions/SEMechanicalVentilation.h>

#include "io/cdm/PatientActions.h"
#include <biogears/cdm/Serializer.h>
#include <biogears/cdm/properties/SEScalarFraction.h>
#include <biogears/cdm/properties/SEScalarPressure.h>
#include <biogears/cdm/properties/SEScalarVolumePerTime.h>
#include <biogears/cdm/substance/SESubstance.h>
#include <biogears/cdm/substance/SESubstanceFraction.h>
#include <biogears/cdm/substance/SESubstanceManager.h>
#include <biogears/schema/cdm/Properties.hxx>


namespace biogears {
SEMechanicalVentilation::SEMechanicalVentilation()
  : SEPatientAction()
{
  m_State = SEOnOff::Invalid;
  m_Flow = nullptr;
  m_Pressure = nullptr;
}
//-------------------------------------------------------------------------------
SEMechanicalVentilation::~SEMechanicalVentilation()
{
  Clear();
}
//-------------------------------------------------------------------------------
void SEMechanicalVentilation::Clear()
{
  SEPatientAction::Clear();

  m_State = SEOnOff::Invalid;
  SAFE_DELETE(m_Flow);
  SAFE_DELETE(m_Pressure);

  DELETE_VECTOR(m_GasFractions);
  m_cGasFractions.clear();
}
//-------------------------------------------------------------------------------
bool SEMechanicalVentilation::IsValid() const
{
  if (!HasState()) {
    Error("Mechanical Ventilation must have state.");
    return false;
  }
  if (GetState() == SEOnOff::Off)
    return true;
  else {
    if (HasGasFraction()) {
      double total = 0;
      for (const SESubstanceFraction* sf : m_cGasFractions) {
        total += sf->GetFractionAmount();
      }
      if (!SEScalar::IsValue(1, total)) {
        Error("Mechanical Ventilation Gas fractions do not sum to 1");
        return false;
      }
    }
  }
  if (!HasPressure() && !HasFlow()) {
    Error("Mechanical Ventilation must have pressure and/or flow");
    return false;
  }
  return true;
}
//-------------------------------------------------------------------------------
bool SEMechanicalVentilation::IsActive() const
{
  if (!HasState())
    return false;
  return GetState() == SEOnOff::On;
}
//-------------------------------------------------------------------------------
SEOnOff SEMechanicalVentilation::GetState() const
{
  return m_State;
}
void SEMechanicalVentilation::SetState(SEOnOff state)
{
  m_State = state;
}
bool SEMechanicalVentilation::HasState() const
{
  return m_State == SEOnOff::Invalid ? false : true;
}
void SEMechanicalVentilation::InvalidateState()
{
  m_State = SEOnOff::Invalid;
}
//-------------------------------------------------------------------------------
bool SEMechanicalVentilation::HasFlow() const
{
  return m_Flow == nullptr ? false : m_Flow->IsValid();
}
SEScalarVolumePerTime& SEMechanicalVentilation::GetFlow()
{
  if (m_Flow == nullptr)
    m_Flow = new SEScalarVolumePerTime();
  return *m_Flow;
}
double SEMechanicalVentilation::GetFlow(const VolumePerTimeUnit& unit) const
{
  if (m_Flow == nullptr)
    return SEScalar::dNaN();
  return m_Flow->GetValue(unit);
}
//-------------------------------------------------------------------------------
bool SEMechanicalVentilation::HasPressure() const
{
  return m_Pressure == nullptr ? false : m_Pressure->IsValid();
}
SEScalarPressure& SEMechanicalVentilation::GetPressure()
{
  if (m_Pressure == nullptr)
    m_Pressure = new SEScalarPressure();
  return *m_Pressure;
}
double SEMechanicalVentilation::GetPressure(const PressureUnit& unit) const
{
  if (m_Pressure == nullptr)
    return SEScalar::dNaN();
  return m_Pressure->GetValue(unit);
}
//-------------------------------------------------------------------------------
bool SEMechanicalVentilation::HasGasFraction() const
{
  return m_GasFractions.empty()  ? false : true;
}
bool SEMechanicalVentilation::HasGasFraction(const SESubstance& s) const
{
  for (const SESubstanceFraction* sf : m_GasFractions) {
    if (&s == &sf->GetSubstance())
      return true;
  }
  return false;
}
const std::vector<SESubstanceFraction*>& SEMechanicalVentilation::GetGasFractions()
{
  return m_GasFractions;
}
const std::vector<const SESubstanceFraction*>& SEMechanicalVentilation::GetGasFractions() const
{
  return m_cGasFractions;
}
SESubstanceFraction& SEMechanicalVentilation::GetGasFraction(SESubstance& s)
{
  for (SESubstanceFraction* sf : m_GasFractions) {
    if (&s == &sf->GetSubstance())
      return *sf;
  }
  SESubstanceFraction* sf = new SESubstanceFraction(s);
  sf->GetFractionAmount().SetValue(0);
  m_GasFractions.push_back(sf);
  m_cGasFractions.push_back(sf);
  return *sf;
}
const SESubstanceFraction* SEMechanicalVentilation::GetGasFraction(const SESubstance& s) const
{
  const SESubstanceFraction* sf = nullptr;
  for (unsigned int i = 0; i < m_GasFractions.size(); i++) {
    sf = m_GasFractions[i];
    if (&s == &sf->GetSubstance())
      return sf;
  }
  return sf;
}
void SEMechanicalVentilation::RemoveGasFraction(const SESubstance& s)
{
  const SESubstanceFraction* sf;
  for (unsigned int i = 0; i < m_GasFractions.size(); i++) {
    sf = m_GasFractions[i];
    if (&s == &sf->GetSubstance()) {
      m_GasFractions.erase(m_GasFractions.begin() + i);
      m_cGasFractions.erase(m_cGasFractions.begin() + i);
      delete sf;
    }
  }
}
void SEMechanicalVentilation::RemoveGasFractions()
{
  DELETE_VECTOR(m_GasFractions);
  m_cGasFractions.clear();
}
//-------------------------------------------------------------------------------
void SEMechanicalVentilation::ToString(std::ostream& str) const
{
  str << "Patient Action : Mechanical Ventilation";
  if (HasComment())
    str << "\n\tComment: " << m_Comment;

  str << "\n\tState: ";
  HasState() ? str << GetState() : str << "Not Set";
  str << "\n\tFlow: ";
  HasFlow() ? str << *m_Flow : str << "Not Set";
  str << "\n\tPressure: ";
  HasPressure() ? str << *m_Pressure : str << "Not Set";
  if (HasGasFraction()) {
    for (const SESubstanceFraction* sf : GetGasFractions()) {
      str << "\n\tSubstance : " << sf->GetSubstance().GetName() << " Fraction Amount " << sf->GetFractionAmount();
    }
  }
  str << std::flush;
}
//-------------------------------------------------------------------------------
bool SEMechanicalVentilation::operator==(const SEMechanicalVentilation& rhs) const
{
  /*
  
  SEOnOff m_State;
  SEScalarVolumePerTime* m_Flow;
  SEScalarPressure* m_Pressure;

  std::vector<SESubstanceFraction*> m_GasFractions;
  std::vector<const SESubstanceFraction*> m_cGasFractions;
  */

  bool equivilant = m_Comment == rhs.m_Comment;
  equivilant &= m_State == rhs.m_State;
  equivilant &= (m_Flow && rhs.m_Flow) ? m_Flow->operator==(*rhs.m_Flow) : m_Flow == rhs.m_Flow;
  equivilant &= (m_Pressure && rhs.m_Pressure) ? m_Pressure->operator==(*rhs.m_Pressure) : m_Pressure == rhs.m_Pressure;
  equivilant &= m_GasFractions.size() == rhs.m_GasFractions.size();
  if (equivilant) {
    for (auto i = 0; i < m_GasFractions.size(); ++i) {
      equivilant &= (m_GasFractions[i] && rhs.m_GasFractions[i]) ? m_GasFractions[i]->operator==(*rhs.m_GasFractions[i]) : m_GasFractions[i] == rhs.m_GasFractions[i];
    }
  }
  equivilant &= m_GasFractions.size() == rhs.m_GasFractions.size();

  return equivilant;
}
//-------------------------------------------------------------------------------
bool SEMechanicalVentilation::operator!=(const SEMechanicalVentilation& rhs) const
{
  return !(*this == rhs);
}
}